### DISCUSSION TITLE
Java Central Station

### DISCUSSION THREAD

**Discussion by fransjacobs** [fransjacobs/model-railway#21] [General]
Java Central Station, or JCS is a project to control a model railroad. I started this project as a hobby back in 2019 and with ups and down it is slowly progressing. Aim of the project is to have fun an to learn something about Model Train control.

I first started this to show on a bigger screen what going on on my layout as the Maerklin CS 2/3 have a tiny screen.
This base functionality is still working but slowly the idea is to also automate the running of the trains.

So the following features are working:

- Communication with the CS2/3.
- Edit and display the layout.
- The Layout screen shows the Turnouts and signals with the current value (The direction of the turnout is shown in Green or Red).
- The Sensors are shows etc.
- You can switch a Turnout from either the CS 2/3 or the screen
- You can control a locomotive (speed and functions)
- Track Voltage and  Current and CS temperature are shown

Next feature is automatic control.
So for that I have added routing of the track so it will now show a route from block to block.
Next question is then how to setup the automatic control? 
This is a topic which is really brainteaser.
I have not yet figured out how this should or could work, so ideas are welcome ;)

**Comment by alma554** [comment]
Hello, I came across your model railway software on github today. I find this very interesting and have downloaded the program. The only thing I can't do is install it successfully. It always shows me that the application is a Java Runtime Environment. What am I doing wrong. Java is installed.

**Reply by fransjacobs** [reply]
Hi alma554,
Thank you for trying my project.

I find it hard to give you a solution as there are so many things which can go wrong.
Some questions:

- on which OS are you trying to install?
- which JDK is installed (only 21 is supported). 
- Is this JDK on the path (try java -version in a command shell)
- Did you follow the [wiki pages](https://github.com/fransjacobs/model-railway/wiki/Installing-and-preparing-JCS)?
- Did you try to run in from a command shell? What error does it then give?

Regards,

Frans

**Comment by alma554** [comment]
Hello Frans,
thanks for the quick reply. I didn't have the correct java version installed. Now it works and I can try your program.
Thank you
Eddie

**Comment by alma554** [comment]
Hello Frans,
I tested your JCS today. First of all, big praise for this easy-to-use program. A lot of things worked well. Creating a track plan, connecting to my DCC-EX CS and also creating and driving locomotives. Unfortunately I cannot switch points and the sensors are not activated. I assume that it is because I control the switch sevs with the PCA 9685 servo driver module and also connected the sensors directly to the DCC-EX CS. This option is described on the DCC-EX page and I saw it as an advantage because you don't need an extra decoder. My question would be whether I'm doing something wrong or whether you can add something to the program to take advantage of this control option.
Have a nice weekend
Eddie

**Reply by fransjacobs** [reply]
Hi Eddie,
Thanks for Testing!

The direct drive of servo's and direct sensor read out on the DCC-EX is not yet supported.
I use an external DDC decoder for my points and an external Sensor decoder (HSI-S88) for reading out sensors.

I have added issues [59](https://github.com/fransjacobs/model-railway/issues/59) and [60](https://github.com/fransjacobs/model-railway/issues/60) to add these features.

Regards,

Frans
